### PR TITLE
Add examples of using reproject with solar data

### LINKED ISSUE #3209: Add a reproject example to the gallery
The [reproject](https://github.com/astrofrog/reproject) package works with solar data without too much trickery we should demo it in the gallery. An example of what you can do is: https://gist.github.com/Cadair/3c95db6dd5e723b3f09b9f495eec6546

To Do for this:

- Make sure there is a bug health warning on the example because it's not flux conserving. (it interpolates)
- Work out a better way of updating the observer location in the header.

### LINKED ISSUE #3353: Add a mosaicing example to the gallery
Now astropy/reproject#186 is merged, I tested it out with AIA and EUVI data. The result was awesome: https://nbviewer.jupyter.org/gist/Cadair/3e96550441433c7d7d4219f076d8b558

At some point that should be put into the example gallery.

It requires #3350 and also code currently in astropy/reproject#52

### PR DESCRIPTION
**Pull request by dstansby**
This is an experimental attempt to add an example demonstrating how to take a standard image of the Sun taken from e.g. Earth and transform it into a Carrington frame using `reproject`. Very open to discussion including whether this belongs in the SunPy examples gallery.

closes #3353 
closes #3209

TODO:
- [ ] Check the rendered results are nice

### CODE REVIEW

**Comment by Cadair** [comment]
I definately want to see a few reprojection examples in the gallery before 1.1, this is an excellent start.

I have spoken to @astrofrog about hopefully getting a release of reproject out before our 1.1 release, I don't really want to merge these examples without knowing that is on the way though.

Would you mind if we turned this into a PR for a bunch of these? I have that mosaicing one and there is the STEREO <> AIA HPC transform example as well. I think having a few where they all talk about some of the steps in reprojection would be good?

Do we want a section specifically for reproject? or shall we have a section for using other packages with SunPy? We could put the examples that use astroquery in a subsection of that heading as well?

File: examples/map/map_carrington_reprojection.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+aia_map = sunpy.map.Map(sunpy.data.sample.AIA_193_IMAGE)
+
+fig = plt.figure()
+ax = plt.subplot(projection=aia_map)
+aia_map.plot(ax)
+
+###############################################################################
+# Now construct a header for the output
+shape_out = [720, 1440]
+header = sunpy.map.make_fitswcs_header(np.empty(shape_out),
```
Comment (Cadair) [inline examples/map/map_carrington_reprojection.py]: Just a general thing, I really don't like how we have to allocate a whole array just for this.

**Comment by dstansby** [comment]
Yes, definitely feel free to hijak this to add more! I think a section on "Other packages" would be good, I was struggling on whether to put this in "Maps" or "Coordinates", but I think a new section sounds best.

File: changelog/3405.feature.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+`sunpy.map.make_fitswcs_header` now accepts a `tuple` representing the shape of an array as well as the actual array as the ``data`` argument.
```
Comment (dstansby) [inline changelog/3405.feature.rst]: Please can this go in another PR? This one is just for the example!

**Comment by Cadair** [comment]
I would like to add one more example to this, an HPC observer shift example.

**Comment by Cadair** [comment]
I have now added all the different examples I can think of to this. :grin:

File: sunpy/map/header_helper.py [hunk 3]
```diff
@@ -49,8 +49,8 @@ def make_fitswcs_header(data, coordinate,
 
     Parameters
     ----------
-    data : `~numpy.ndarray`
-        Array data of Map for which a header is required.
+    data : `~numpy.ndarray` or `tuple`
+        Array data of Map for which a header is required, or the shape of an array.
```
Comment (hayesla) [inline sunpy/map/header_helper.py]: should we be more explicit here to the user about the shape - i.e. that (input1, input2) is (y_size, x_size) of array - I know that this is the way it should be as row major in Python but people might just get confused when thinking in terms of a map?

File: examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+######################################################################
+# Next we create a SunPy map for each of the files.
+#
+maps = sunpy.map.Map(sorted(files))
+
+
+######################################################################
+# When combining these images all three need to assume the same radius of
+# the Sun for the data. The AIA images specify a slightly different value
+# that the IAU 2015 constant. To avoid coordinate transformation issues we
```
Comment (hayesla) [inline examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py]: that -> than

File: examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+ax.set_rlim(0, 1.3)
+
+plt.legend()
+plt.show()
+
+
+######################################################################
+# The next step is to calculate the output coordinate system for the
+# combined map. We select a heliographic Stonyhurst frame, and a Plate
+# Caree projection, and generate a header using
```
Comment (hayesla) [inline examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py]: Plate Carrée (CAR)

File: examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+######################################################################
+# Improving the Output
+# --------------------
+#
+# As you can see this leaves a little to be desired. To reduce the obvious
+# warping towards the points which are close to the limb in the input
+# images, we can define a set of weights to use when coadding the output
+# arrays. To reduce this warping we want to calculate an set of weights
+# which weights highly points close to the centre of the disk in the input
```
Comment (hayesla) [inline examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py]: ..._a_ set of weights which highly weigh points ...

**Comment by Cadair** [comment]
I have now rebased this ontop of #3483 which contains the change to `make_fitswcs_header`.

**Comment by Cadair** [comment]
> I'm not sure that "Using SunPy with Other Packages" accomplishes that; maybe we can add more meaningful section titles like: "How to identify other astronomical objects" and "How to co-align image data".

My concern with this is that we would end up with a lot of sections, one for each third party package we want examples with. I wonder if we could have subsections or something.

Thanks for the other review comments, I will try and address them. It should be noted that the two astroquery examples I only moved and not added :grin:

**Comment by mbobra** [comment]
> My concern with this is that we would end up with a lot of sections, one for each third party package we want examples with. I wonder if we could have subsections or something.

Yeah, totally I agree that sub-subsections would be ideal (e.g. Example Gallery -> Units and Coordinates -> some sub-subsection). Unfortunately, the [Sphinx-Gallery documentation](https://sphinx-gallery.github.io/configuration.html#manage-multiple-galleries) says: 

> Sphinx-Gallery only supports up to sub-folder level in its gallery directories.

Maybe we can organize these 8 examples this way:

Under **Units and Coordinates**:
* `STEREO_SECCHI_starfield.py`
* `getting_lasco_observer_location.py`
* `AIA_limb_STEREO.py`
* `SDO_to_STEREO_Coordinate_Conversion.py`

Create a new section called **Combining, Co-aligning, and Reprojecting Image Data** and add these examples:
* `reprojection_aia_euvi_mosaic.py`
* `reprojection_align_aia_hmi.py`
* `reprojection_different_observers.py`
* `reprojection_heliographic_stonyhurst.py`

**Comment by mbobra** [comment]
I can create a template for the gallery examples and write an introduction for all 8 of these examples. I'll do it at the end of this week and will update this PR on Thursday or Friday.

File: examples/sunpy_other_packages/reprojection_different_observers.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+files = Fido.fetch(res)
+
+
+######################################################################
+# Create a map for each image.
+#
+
+map_aia, map_stereo = sunpy.map.Map(sorted(files))
+
+fig = plt.figure()
```
Comment (dstansby) [inline examples/sunpy_other_packages/reprojection_different_observers.py]: Just `peek()` these figures?

File: examples/sunpy_other_packages/reprojection_different_observers.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    map_aia.reference_coordinate,
+    scale=u.Quantity(map_aia.scale)*4,
+    instrument="EUVI",
+    observatory="AIA Observer",
+    wavelength=map_stereo.wavelength
+)
+
+
+######################################################################
+# Next we construct an ``~astropy.wcs.WCS`` object from the header.
```
Comment (dstansby) [inline examples/sunpy_other_packages/reprojection_different_observers.py]:
```suggestion
# Next we construct an `~astropy.wcs.WCS` object from the header.
```

File: examples/sunpy_other_packages/reprojection_different_observers.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+# Currently ``~astropy.wcs.WCS`` does not understand the observer
```
Comment (dstansby) [inline examples/sunpy_other_packages/reprojection_different_observers.py]:
```suggestion
# Currently `~astropy.wcs.WCS` does not understand the observer
```

File: examples/sunpy_other_packages/reprojection_different_observers.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+# position, so we manually set that.
+#
+
+out_wcs = WCS(out_header)
+out_wcs.heliographic_observer = map_aia.reference_coordinate.observer
+
+
+######################################################################
+# We can now reproject the stereo map to this output ``~astropy.wcs.WCS``.
```
Comment (dstansby) [inline examples/sunpy_other_packages/reprojection_different_observers.py]:
```suggestion
# We can now reproject the STEREO map to this output `~astropy.wcs.WCS`.
```

File: examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+#
+
+earth = get_body_heliographic_stonyhurst('earth', maps[0].date)
+
+plt.figure(figsize=(8, 8))
+r_unit = u.AU
+
+ax = plt.subplot(projection='polar')
+circle = plt.Circle((0.0, 0.0), (10*u.Rsun).to_value(r_unit),
+                    transform=ax.transProjectionAffine + ax.transAxes, color="yellow",
```
Comment (dstansby) [inline examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py]: This is a bit complicated, why does it need the transform bit?

**Comment by dstansby** [comment]
Some minor comments, but this is looking ⭐️⭐️⭐️⭐️⭐️ overall

File: examples/sunpy_other_packages/reprojection_align_aia_hmi.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+res = Fido.search(time, aia | hmi)
+
+files = Fido.fetch(res[:, 0])
+
+
+######################################################################
+# Create a map for each image.
+#
+
+map_aia, map_hmi = [m.resample((1024, 1024)*u.pix) for m in sunpy.map.Map(sorted(files))]
```
[the remark this answers is not in the document]
Comment (Cadair) [reply examples/sunpy_other_packages/reprojection_align_aia_hmi.py]: Just to reduce the computation time for the docs build. There should indeed be a comment to that effect.

File: examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py (hunk 11 above)
Comment (Cadair) [reply examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py]: I copied this out of @hayesla 's other example of plotting all the planet positions, so I am not 100% sure!

File: examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+import sunpy.data.sample
+import sunpy.map
+
+###############################################################################
+# We start with the sample data
+aia_map = sunpy.map.Map(sunpy.data.sample.AIA_193_IMAGE)
+
+fig = plt.figure()
+ax = plt.subplot(projection=aia_map)
+aia_map.plot(ax)
```
[the remark this answers is not in the document]
Comment (Cadair) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: I have a suspicion that we have a policy of not using `peek` in examples and stuff. I can't remember exactly where we ended up on this. Maybe @ayshih has thoughts.

**Comment by Cadair** [comment]
@dstansby As it's your PR can you accept all your own suggestions because I can't lol.

File: examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ayshih) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: Heh, of course I have thoughts!  Yes, we essentially have an unwritten policy of not using `peek()` in examples when the objective is to create a plot, because that's what `plot()` is for.  The exact nature of `peek()` – its use cases, what it produces, and even its name – is still subject to discussion and change, and so `peek()` should not be relied on to create plots.

File: examples/sunpy_other_packages/reprojection_different_observers.py (hunk 7 above)
Comment (ayshih) [reply examples/sunpy_other_packages/reprojection_different_observers.py]: See other comment

File: examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+######################################################################
+# To display the output we construct a new map using the new array and our
+# generated header. We also borrow the plot settings from the AIA map.
+#
+
+outmap = sunpy.map.Map((array, header))
+outmap.plot_settings = maps[0].plot_settings
+
+outmap.peek()
```
Comment (ayshih) [inline examples/sunpy_other_packages/reprojection_aia_euvi_mosaic.py]: See comment about not using `peek()`

**Comment by wafels** [comment]
This needs reproject > version 0.6, right?  I may have missed this in the text of the PR, but this requirement should be in here.

File: examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dstansby) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: But isn't the whole point of `peek()` to quickly have a glance at a plot, which is the use case here? If an un-written policy is not to use peek, then it should be a written policy in the `peek()` docs...

[the remark this answers is not in the document]
Comment (ayshih) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: Not exactly.  You may recall past discussions on `peek()` (e.g., https://github.com/sunpy/sunpy/pull/3103#issuecomment-494569464).  Here's what the docstring for Map's `peek()` says:

> Displays a graphical overview of the data in this object for user evaluation. For the creation of plots, users should instead use the `plot` method and Matplotlib’s pyplot framework.

While this "graphical overview" does (currently) take the form of a Matplotlib plot, we don't want to inadvertently teach people to use `peek()` instead of using `plot()` when creating plots, so that's why we want to refrain from using `peek()` in examples.

`peek()` is great – I use it all the time – but it needs to evolve into its own thing beyond just a shorthand for `plot()`.

[the remark this answers is not in the document]
Comment (dstansby) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: I'm still a bit confused - what's the difference between a 'graphical overview' and a 'plot'?

[the remark this answers is not in the document]
Comment (mbobra) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: I'm trying to update the [example gallery section of the developer's guide](https://docs.sunpy.org/en/latest/dev_guide/example_gallery.html#) and [example template](https://github.com/sunpy/sunpy/blob/master/examples/example_template/example_template.py). Are there any other unofficial policies, besides not using `peek()`, that I should include?

[the remark this answers is not in the document]
Comment (ayshih) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: > I'm still a bit confused - what's the difference between a 'graphical overview' and a 'plot'?

Think of the difference between the text `__repr__()` of a large `ndarray` and converting the `ndarray` contents to a string.  I argue that `peek()` should be a graphical analogue to `__repr__()`.  In my mind, it'd be fine if the "plot" aspect of `peek()` were to be diminished or abbreviated.  There's more to a `Map` object than how it plots; for example, the graphical overview can/should include important metadata or data characteristics.  If you want to see how the object plots, use `plot()`.

> Are there any other unofficial policies, besides not using `peek()`, that I should include?

Not that I'm aware of.  If/when `peek()` evolves into something more than it is, I think it'd be totally fine to include in examples (in the same manner that we show text `__repr__()` output).  It's just that while it currently produces output that is just like `plot()` output, people will (incorrectly) learn it as a shorthand to produce plots.

[the remark this answers is not in the document]
Comment (dstansby) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: > 
> Think of the difference between the text `__repr__()` of a large `ndarray` and converting the `ndarray` contents to a string. I argue that `peek()` should be a graphical analogue to `__repr__()`. In my mind, it'd be fine if the "plot" aspect of `peek()` were to be diminished or abbreviated. There's more to a `Map` object than how it plots; for example, the graphical overview can/should include important metadata or data characteristics. If you want to see how the object plots, use `plot()`.

Okay, but showing the image as part of `peek()` is presumably never going to go away? The purpose of this bit of code in the example is to take a quick look at what's in the `Map` object, not to create a scientific plot, so I'd argue that `peek()` is definitely appropriate in this place in the example.

[the remark this answers is not in the document]
Comment (ayshih) [reply examples/sunpy_other_packages/reprojection_heliographic_stonyhurst.py]: I agree that if `peek()` were already doing something tangibly different from `plot()`, it'd be the appropriate method to call here.  Given its current output, though, we know that users will get the wrong idea and learn it as a shorthand for `plot()`, and then that will have to be un-learned later.

**Comment by mbobra** [comment]
@dstansby I realized changing the example template is not trivial, and I'd like some more time to think about it. For now, can you re-write the introductions for each example using the following general format? Each introduction doesn't need to be more than 3 sentences.

1. **Give some background information to motivate the example.** E.g. (from `STEREO_SECCHI_starfield.py`), "Since the field of view in SECCHI COR2 images can span 2 to 15 solar radii, we often observe stars in these image data."

2. **State the goal of the example. This should include clear, succinct references to software packages and observational data that the reader can peruse for more information.** E.g. "In this example, we will use the [Astroquery package](https://astroquery.readthedocs.io/en/latest/) to query the [VizieR star catalog](http://vizier.u-strasbg.fr/viz-bin/VizieR) for stars observed by the [Gaia satellite](https://sci.esa.int/web/gaia/) within the SECCHI COR2 field of view."

3. **State the relevant SunPy functionality necessary to accomplish this goal.** E.g. "To accomplish this, will use the coordinates framework in SunPy and AstroPy to transform the coordinates returned by VizieR into SECCHI COR2 image coordinates."

4. **Add any additional information, if desired.** E.g. "As a bonus, we'll also identify Mars."

**Comment by nabobalis** [comment]
There is a copy of the astropy example template in the examples folder, we just don't add it to our gallery.

**Comment by nabobalis** [comment]
I did some tidying up of the files and rebased ontop of master due to merge conflict in the conf.py file.

**Comment by mbobra** [comment]
> There is a copy of the astropy example template in the examples folder, we just don't add it to our gallery.

Yup, I think we can improve on this. I'm working on it 😄
